org.eclipse.m2e.core/src/org/eclipse/m2e/core/embedder/IMavenExecutionContext.java

```diff
@@ -106,4 +125,31 @@ public interface IMavenExecutionContext {
    */
   ProjectBuildingRequest newProjectBuildingRequest();
 
+  /**
+   * Either join the currents thread {@link IMavenExecutionContext} or creates a fresh one if this thread is currently
+   * not executing. Be aware that because of this might return an already executing context, this context can't be
+   * configured in any way, use {@link IMaven#createExecutionContext()} or
+   * {@link IMavenProjectFacade#createExecutionContext()} to create a specific context that could be configured.
+   * 
+   * @return a {@link IMavenExecutionContext} that could be used for execution
+   * @throws CoreException
+   */
+  static IMavenExecutionContext join() throws CoreException {
+    IMavenExecutionContext context = MavenExecutionContext.getThreadContext();
+    if(context == null) {
+      return MavenPlugin.getMaven().createExecutionContext();
```

Comment:
I don't think it's a good thing to have in IMavenExecutionContext, that is supposed to be deduced from one particular IMaven instance, a reference to the common/singleton IMaven. IMO, it's still something that makes more sense to keep on the consumer end or in IMaven.

Reply:
This is (as of now) the usual way for any code that do not want to be specific, to use the global execution as a default.
If the code cares one needs to create a special `IMavenExecutionContext` and not use join.

Reply:
> the usual way for any code that do not want to be specific

Yes, but we already have 1 entry point for the global execution (`MavenPlugin.getMaven()`), I don't think we should create new ones; ideally `MavenExecutionContext` shouldn't reference any `MavenCore.getMaven()` and should only use its own IMaven reference. Having consumers invoking `MavenPlugin.getMaven().getExecutionContext()` (was `context()` would IMO be cleaner) as it forces user to care about choosing IMaven they want to use to execute.

Reply:
> Yes, but we already have 1 entry point for the global execution (`MavenPlugin.getMaven()`)

That's exactly what I want to get rid of.

> ideally `MavenExecutionContext` shouldn't reference any `MavenCore.getMaven()`

sure, actually its not the `MavenExecutionContext` (object) but a static method I just feel its best placed in the interface as in a separate class/util method.

Reply:
> That's exactly what I want to get rid of.

We won't be able to remove `MavenPlugin.getMaven()` for a long time, and in any case, we'll probably still need it forever for compatibility and because we do need an access to default IMaven.
Replacing 1 static access by another doesn't change much IMO, and this hides to users some important coupling without removing this coupling. I don't get the benefit of not having it a (non-static) method in IMaven.

Reply:
> We won't be able to remove `MavenPlugin.getMaven()` for a long time, and in any case, we'll probably still need it forever for compatibility and because we do need an access to default IMaven.

I don't want to get rid of the `MavenPlugin.getMaven()`, but about IMaven#execute(....) methods! That's why they are marked for removal now, the context where we Execute is `IMavenExecutionContext`

> I don't get the benefit of not having it a (non-static) method in IMaven.

because it has nothing to do with `IMaven`, it is just a fallback for the case there is no active executing method, and we need this for two cases:

1. I want to execute something, I don't know (or care) about `IMaven` but want to join if the current thread already has a "transaction"
2. for backward compatibility reasons and we don't want the user to check for null all the time.

Reply:
I have now adjusted to not create an execution implicitly but forcing any code that uses join to run inside a execute code block.